File: nim/day12.nim
import strutils, strscans

let input = "./inputs/12.txt"

const
  length = 300
  leftPad = 50
var
  state = ".".repeat(length)
  fruitful: seq[string]

for line in input.lines:
  var s: string
  if line.scanf("initial state:"):
    let initial = line.split[^1]
    for i in 0 ..< initial.len:
      state[leftPad + i] = initial[i]
  elif line.scanf("$+ => #", s):
    fruitful.add s


type PlantCount = tuple[plants, total: int]

func countPlants(state: string): PlantCount =
  for i, c in state:
    if c == '#':
      result.plants += 1
      result.total += i - leftPad

proc live(state: string, generations: int64): int64 =
  var
    previous: PlantCount
    state = state
    newState: string
  for generation in 1 .. generations:
    newState = ".".repeat(length)
    for i in 2 ..< length-2:
      if state[i-2 .. i+2] in fruitful:
        newState[i] = '#'
    let (plants, total) = countPlants newState
    if plants == previous.plants and total == previous.total + plants:
      let remaining = generations - generation
      return total + plants * remaining
    previous = (plants, total)
    state = newState
  return previous.total


echo state.live(20)
echo state.live(50_000_000_000)

File: python/day12.py
from aoc import *


def grow(plants):
  creates_new_life = lambda i: cat('.#'[j in plants] for j in range(i-2, i+3)) in fruitful
  return [i for i in range(plants[0]-2, plants[-1]+3) if creates_new_life(i)]

def live(plants, generations=20):
  new_plants = plants[:]
  for _ in range(generations):
    new_plants = grow(new_plants)
  return new_plants

def part_2(plants, generations=50_000_000_000):
  after_200 = live(plants, 200)
  increase_rate = sum(grow(after_200)) - sum(after_200)
  return sum(after_200) + (generations - 200) * increase_rate


data = read_input(12)

initial = [i for i, c in enumerate(data[0][15:]) if c == '#']
fruitful = [(l := line.split())[0] for line in data[2:] if line[-1] == '#']

print(sum(live(initial)))
print(part_2(initial))
